File: firmware/stick/src/sap_flow.cpp
/*
 * DIFFUSER STICK — STANDALONE SAP FLOW
 *
 * Ported from tree SapFlowForeground. Particles rise along
 * the 23-LED stick on a deep green background.
 * Runs entirely on the ESP32 — no streaming required.
 */

#include <Adafruit_NeoPixel.h>

#define TOTAL_LEDS (STRIP_A_COUNT + STRIP_B_COUNT)

Adafruit_NeoPixel stripA(STRIP_A_COUNT, STRIP_A_PIN, NEO_GRB + NEO_KHZ800);
Adafruit_NeoPixel stripB(STRIP_B_COUNT, STRIP_B_PIN, NEO_GRB + NEO_KHZ800);

// Frame buffer: R, G, B per LED
uint8_t leds[TOTAL_LEDS][3];

// --- Sap flow parameters (from tree SapFlowForeground.h) ---

struct SapParticle {
    float pos;
    float velocity;
    uint8_t brightness;
    bool active;
};

static const uint8_t MAX_PARTICLES = 12;
SapParticle particles[MAX_PARTICLES];

// Particle color: bright lime green
static const uint8_t SAP_R = 100;
static const uint8_t SAP_G = 255;
static const uint8_t SAP_B = 60;

// Background: deep forest green (visible but subtle)
static const uint8_t BG_R = 1;
static const uint8_t BG_G = 6;
static const uint8_t BG_B = 1;

// Spawn timing (tuned for 0.3 velocity)
static const uint8_t SPAWN_CHANCE = 5;         // % per eligible frame
static const uint8_t MIN_FRAMES_BETWEEN = 27;
static const uint8_t MAX_FRAMES_BETWEEN = 160;

// Rendering
static const float VELOCITY = 0.3;
static const float SOFT_RADIUS = 3.0;

uint8_t framesSinceLastSpawn = 0;

// Target ~60 fps
static const unsigned long FRAME_MS = 16;

void showFrame() {
    for (uint16_t i = 0; i < STRIP_A_COUNT; i++)
        stripA.setPixelColor(i, leds[i][0], leds[i][1], leds[i][2]);
    for (uint16_t i = 0; i < STRIP_B_COUNT; i++)
        stripB.setPixelColor(i, leds[STRIP_A_COUNT + i][0], leds[STRIP_A_COUNT + i][1], leds[STRIP_A_COUNT + i][2]);
    stripA.show();
    stripB.show();
}

void setup() {
    Serial.begin(115200);

    stripA.begin();
    stripA.setBrightness(255);
    stripB.begin();
    stripB.setBrightness(255);

    // Startup flash: brief red then off
    for (uint16_t i = 0; i < STRIP_A_COUNT; i++) stripA.setPixelColor(i, 50, 0, 0);
    for (uint16_t i = 0; i < STRIP_B_COUNT; i++) stripB.setPixelColor(i, 50, 0, 0);
    stripA.show();
    stripB.show();
    delay(200);
    stripA.clear(); stripA.show();
    stripB.clear(); stripB.show();

    for (uint8_t i = 0; i < MAX_PARTICLES; i++)
        particles[i].active = false;

    Serial.println("Diffuser Stick — Sap Flow");
    Serial.printf("Strip A: pin %d, %d LEDs\n", STRIP_A_PIN, STRIP_A_COUNT);
    Serial.printf("Strip B: pin %d, %d LEDs\n", STRIP_B_PIN, STRIP_B_COUNT);
}

void loop() {
    unsigned long frameStart = millis();

    framesSinceLastSpawn++;

    // --- Spawn ---
    bool shouldSpawn = false;
    if (framesSinceLastSpawn >= MAX_FRAMES_BETWEEN) {
        shouldSpawn = true;
    } else if (framesSinceLastSpawn >= MIN_FRAMES_BETWEEN && random(100) < SPAWN_CHANCE) {
        shouldSpawn = true;
    }

    if (shouldSpawn) {
        for (uint8_t i = 0; i < MAX_PARTICLES; i++) {
            if (!particles[i].active) {
                particles[i].active = true;
                particles[i].pos = 0;
                particles[i].velocity = VELOCITY;
                particles[i].brightness = 80 + random(70);
                framesSinceLastSpawn = 0;
                break;
            }
        }
    }

    // --- Update ---
    for (uint8_t i = 0; i < MAX_PARTICLES; i++) {
        if (!particles[i].active) continue;
        particles[i].pos += particles[i].velocity;
        if (particles[i].pos > TOTAL_LEDS + 5)
            particles[i].active = false;
    }

    // --- Render ---
    // Fill background
    for (uint8_t i = 0; i < TOTAL_LEDS; i++) {
        leds[i][0] = BG_R;
        leds[i][1] = BG_G;
        leds[i][2] = BG_B;
    }

    // Additive particle rendering with soft edges
    for (uint8_t p = 0; p < MAX_PARTICLES; p++) {
        if (!particles[p].active) continue;

        float particlePos = particles[p].pos;
        uint8_t brightness = particles[p].brightness;

        for (uint8_t i = 0; i < TOTAL_LEDS; i++) {
            float distance = fabs((float)i - particlePos);
            if (distance < SOFT_RADIUS) {
                float falloff = 1.0f - (distance / SOFT_RADIUS);
                falloff = falloff * falloff;

                uint8_t pr = (uint16_t)SAP_R * brightness * falloff / 255;
                uint8_t pg = (uint16_t)SAP_G * brightness * falloff / 255;
                uint8_t pb = (uint16_t)SAP_B * brightness * falloff / 255;

                leds[i][0] = min(255, leds[i][0] + pr);
                leds[i][1] = min(255, leds[i][1] + pg);
                leds[i][2] = min(255, leds[i][2] + pb);
            }
        }
    }

    showFrame();

    // Hold frame rate
    unsigned long elapsed = millis() - frameStart;
    if (elapsed < FRAME_MS)
        delay(FRAME_MS - elapsed);
}

File: audio-reactive/jam/client/src/effects/sapflow.ts
import { BaseEffect, type EffectConfig } from './base';
import { computeLedPositions } from '../topology/loader';

/**
 * Sap Flow: particles rise from roots through trunk to canopy tips.
 * Port of SapFlowForeground.h — stochastic particle spawn,
 * constant velocity, soft Gaussian falloff, additive blending.
 * Uses globalProgress (physical height) as the flow axis.
 */

interface Particle {
  height: number;    // 0 (roots) → 1 (tips)
  velocity: number;  // units per second
  brightness: number; // 0-1
  active: boolean;
}

const MAX_PARTICLES = 16;
const VELOCITY = 0.12;          // full tree height in ~8 seconds
const MIN_SPAWN_INTERVAL = 0.4; // seconds
const MAX_SPAWN_INTERVAL = 2.5; // force spawn
const SPAWN_CHANCE = 0.08;      // per frame when eligible
const SOFT_RADIUS = 0.04;       // falloff distance in height units

// Deep forest green background (~1%)
const BG_R = 1;
const BG_G = 6;
const BG_B = 1;

// Sap color: bright lime green
const SAP_R = 100;
const SAP_G = 255;
const SAP_B = 100;

export class SapFlowEffect extends BaseEffect {
  private buf: Uint8Array;
  private heights: Float32Array; // globalProgress per LED
  private particles: Particle[] = [];
  private timeSinceSpawn = 0;

  constructor(config: EffectConfig) {
    super(config);
    this.buf = new Uint8Array(this.numLeds * 3);
    this.heights = new Float32Array(this.numLeds);

    if (config.topology) {
      const computed = computeLedPositions(config.topology);
      for (let i = 0; i < this.numLeds; i++) {
        this.heights[i] = computed.globalProgress[i];
      }
    } else {
      // Fallback: linear mapping
      for (let i = 0; i < this.numLeds; i++) {
        this.heights[i] = i / (this.numLeds - 1);
      }
    }

    for (let i = 0; i < MAX_PARTICLES; i++) {
      this.particles.push({ height: 0, velocity: 0, brightness: 0, active: false });
    }
  }

  processAudio(_freqData: Float32Array, _timeData: Float32Array): void {
    // Non-interactive — ignores audio
  }

  render(dt: number): Uint8Array {
    this.timeSinceSpawn += dt;

    // Spawn logic
    let shouldSpawn = false;
    if (this.timeSinceSpawn >= MAX_SPAWN_INTERVAL) {
      shouldSpawn = true;
    } else if (this.timeSinceSpawn >= MIN_SPAWN_INTERVAL && Math.random() < SPAWN_CHANCE) {
      shouldSpawn = true;
    }

    if (shouldSpawn) {
      for (const p of this.particles) {
        if (!p.active) {
          p.active = true;
          p.height = 0;
          p.velocity = VELOCITY * (0.8 + Math.random() * 0.4); // slight variation
          p.brightness = 0.3 + Math.random() * 0.7;
          this.timeSinceSpawn = 0;
          break;
        }
      }
    }

    // Update particles
    for (const p of this.particles) {
      if (!p.active) continue;
      p.height += p.velocity * dt;
      if (p.height > 1.0 + SOFT_RADIUS * 2) {
        p.active = false;
      }
    }

    // Render: background + additive particles
    for (let i = 0; i < this.numLeds; i++) {
      const off = i * 3;
      let r = BG_R;
      let g = BG_G;
      let b = BG_B;

      const h = this.heights[i];

      for (const p of this.particles) {
        if (!p.active) continue;
        const dist = Math.abs(h - p.height);
        if (dist >= SOFT_RADIUS) continue;

        const falloff = 1.0 - dist / SOFT_RADIUS;
        const intensity = falloff * falloff * p.brightness;

        r += SAP_R * intensity;
        g += SAP_G * intensity;
        b += SAP_B * intensity;
      }

      this.buf[off] = Math.min(255, r);
      this.buf[off + 1] = Math.min(255, g);
      this.buf[off + 2] = Math.min(255, b);
    }

    return this.buf;
  }

  get name(): string {
    return 'sap flow';
  }
}
